crypto/src/proteus.rs

```diff
@@ -0,0 +1,879 @@
+            };
+            keystore.save(db_session).await?;
+        }
+
+        Ok(())
+    }
+
+    /// Deletes a session in the store
+    pub async fn session_delete(&mut self, keystore: &CryptoKeystore, session_id: &str) -> CryptoResult<()> {
+        if keystore.remove::<ProteusSession, _>(session_id).await.is_ok() {
+            let _ = self.proteus_sessions.remove(session_id);
+        }
+        Ok(())
+    }
+
+    /// Session accessor
+    pub fn session(&mut self, session_id: &str) -> Option<&mut ProteusConversationSession> {
+        self.proteus_sessions.get_mut(session_id)
+    }
+
+    /// Decrypt a proteus message for an already existing session
+    /// Note: This cannot be used for handshake messages, see [ProteusCentral::session_from_message]
+    pub async fn decrypt(
+        &mut self,
+        keystore: &mut CryptoKeystore,
+        session_id: &str,
+        ciphertext: &[u8],
+    ) -> CryptoResult<Vec<u8>> {
+        if let Some(session) = self.proteus_sessions.get_mut(session_id) {
+            session.decrypt(keystore, ciphertext).await
+        } else {
+            Err(CryptoError::ConversationNotFound(session_id.as_bytes().into()))
+        }
+    }
+
+    /// Encrypt a message for a session
+    pub fn encrypt(&mut self, session_id: &str, plaintext: &[u8]) -> CryptoResult<Vec<u8>> {
+        if let Some(session) = self.session(session_id) {
+            session.encrypt(plaintext)
+        } else {
+            Err(CryptoError::ConversationNotFound(session_id.as_bytes().into()))
+        }
+    }
+
+    /// Encrypts a message for a list of sessions
+    /// This is mainly used for conversations with multiple clients, this allows to minimize FFI roundtrips
+    pub fn encrypt_batched(
+        &mut self,
+        sessions: &[impl AsRef<str>],
+        plaintext: &[u8],
+    ) -> CryptoResult<HashMap<String, Vec<u8>>> {
+        let mut acc = HashMap::new();
+        for session_id in sessions {
+            if let Some(session) = self.session(session_id.as_ref()) {
+                acc.insert(session.identifier.clone(), session.encrypt(plaintext)?);
+            }
+        }
+        Ok(acc)
+    }
+
+    /// Generates a new Proteus PreKey, stores it in the keystore and returns a serialized PreKeyBundle to be consumed externally
+    pub async fn new_prekey(&self, id: u16, keystore: &CryptoKeystore) -> CryptoResult<Vec<u8>> {
+        use proteus_wasm::keys::{PreKey, PreKeyId};
+
+        let prekey_id = PreKeyId::new(id);
+        let prekey = PreKey::new(prekey_id);
+        let keystore_prekey = core_crypto_keystore::entities::ProteusPrekey::from_raw(
+            id,
+            prekey.serialise().map_err(ProteusError::from)?,
+        );
+        keystore.save(keystore_prekey).await?;
+
+        let bundle = PreKeyBundle::new(self.proteus_identity.as_ref().public_key.clone(), &prekey);
+        Ok(bundle.serialise().map_err(ProteusError::from)?)
```

Comment:
Prevents having a zombie prekey in the keystore in the event bundle serialization fails
```suggestion
        let bundle = PreKeyBundle::new(self.proteus_identity.as_ref().public_key.clone(), &prekey);
        let bundle = bundle.serialise().map_err(ProteusError::from)?;
        keystore.save(keystore_prekey).await?;
        Ok(bundle)
```